step3: 문자열 계산기 구현

KimKwangYoung opened the pull request:
문자열 계산기 구현해봤습니다!

초반에 어떻게 해야할지 감이 잘 안잡혀서 다른분들 코드도 조금 참조를 하고 했는데 
Splitter에서 공백으로 각 문자를 나누는 과정에서 검증을 하는 코드가 난잡한 느낌이네요 
잘 맞게 한 건지 걱정되네요

리뷰 부탁드립니다!😃

KimKwangYoung committed:
step3: 문자열 계산기 구현

laco-dev commented on app/build.gradle:
```diff
@@ -7,7 +7,7 @@ android {
 
     defaultConfig {
-        applicationId "edu.nextstep.camp.calculator"
+        applicationId "edu.edu.nextstep.camp.calculator"
```
복사하면서 잘못 들어갔을까요? 다른곳도 모두 `edu`가 중복해서 들어갔네요. 😱

laco-dev commented on domain/build.gradle:
```diff
@@ -5,4 +5,12 @@ plugins {
+
+    testImplementation "junit:junit:$junit_version"
+    testImplementation "com.google.truth:truth:$truth_version"
+    testImplementation "io.mockk:mockk:$mockk_version"
```
MVC 미션에서는 Mock을 사용하지 않고 진행합니다.

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,4 +1,28 @@
 package edu.nextstep.camp.calculator.domain
 
 class Calculator {
+    fun evaluate(input: String?):Int {
```
1. `input` 보다는 보다 변수가 명확한 의미를 담도록 만들어 보세요
2. 전체적으로 코드 리포매팅이 필요합니다.

KimKwangYoung replied:
리포매팅이라고 하시면 코드에 전체적인 수정이 필요하다는 말씀이실까요? 
아니면 evaluate에 대한 수정이 필요하다는 말씀이실까요?

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,4 +1,28 @@
 
 class Calculator {
+    fun evaluate(input: String?):Int {
+        if (input.isNullOrBlank()) throw IllegalArgumentException("입력값은 null이거나 빈값일 수 없습니다.")
```
코틀린의 require(...) { ... } 함수를 이용해서 보다 간결하게 작성이 가능합니다.
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,4 +1,28 @@
+        var left = 0
+        var operator: Operator? = null
+
+        for(i in 0 until expressionNodeQueue.size) {
+            when(val node = expressionNodeQueue.poll()) {
+                is Operand -> {
+                    result = operator?.calculate?.let { it(left, node.value) } ?: node.value
+                    left = result
+                }
+                is Operator -> {
+                    operator = node
+                }
+            }
+        }
+
```
입력 값이 적절하게 분할 되었다고 "가정" 하고, 인덱스별로 꺼내서 바로 계산하도록 만들어봐도 좋습니다.
```kotlin
for ( 1부터 끝까지 step 2) {
   expression[index] --> 연산자
   expression[index + 1] --> 피연산자
   계산! // 계산이 성립되지 않으면 오류가 발생함
}

```

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Node.kt:
```diff
@@ -0,0 +1,36 @@
+package edu.nextstep.camp.calculator.domain
+
+interface Node {
+
+}
```
Node의 타입이 명확하게 Operator와 Operand만 가지는 경우라면 sealed class를 이용해 타입을 제한하는 방법도 좋습니다.

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Node.kt:
```diff
@@ -0,0 +1,36 @@
+    companion object {
+        fun getFrom(token: String): Operator {
+            return when(token) {
+                "+" -> PLUS
+                "-" -> MINUS
+                "*" -> MULTIPLY
+                "/" -> DIVIDE
```
각 사칙연산 기호도 Operator가 가지고 있도록 하면 좋을 것 같습니다.
```kotlin
values().find { it.symbol == token } ?: throw ...
```

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Splitter.kt:
```diff
@@ -0,0 +1,49 @@
+import java.util.*
+
+
+class Splitter {
```
`Splitter`에 대한 테스트 코드도 작성 해 보면 좋겠습니다.

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Splitter.kt:
```diff
@@ -0,0 +1,49 @@
+
+
+class Splitter {
+    companion object {
```
코틀린에서는 정적 메소드만 제공하는 경우 동반 객체를 사용하지 않고 object로 만들어서 사용합니다.

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Splitter.kt:
```diff
@@ -0,0 +1,49 @@
+    companion object {
+        private const val CHECKING_START_INDEX = 1
+
+        fun split(expression: String): Queue<Node> {
```
미션에서의 요구사항은 앞에서부터 순서대로 계산을 이어 나가는데, 자료구조를 Queue로 선택한 이유가 있을까요?

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Splitter.kt:
```diff
@@ -0,0 +1,49 @@
+            val nodeQueue: Queue<Node> = LinkedList()
+            nodeQueue.offer(beforeNode)
+
+            for (i in CHECKING_START_INDEX until splitList.size) {
+                val node = toNode(splitList[i])
+                if (node is Operand && beforeNode !is Operator) {
+                    throw IllegalArgumentException("피연산자 이전 항목은 연산자여야만 합니다.")
+                } else if (node is Operator && beforeNode !is Operand) {
+                    throw IllegalArgumentException("연산자 이전 항목은 피연산자이어야만 합니다.")
+                } else {
+                    nodeQueue.offer(node)
+                    beforeNode = node
+                }
+            }
```
요구사항에 대한 디테일을 상당히 신경써서 위와 같은 로직을 작성 해 주신 것으로 보입니다.
그러다보니 로직이 다소 복잡해지고 코드도 어려워질 수 있는데요.

우선은 쉽게 생각해서 계산을 하면서 계산이 논리적으로 성립할 수 없다면 `IllegalArgumentException`을 던지도록 단순화 해 보면 어떨까요?

```
"2 + + 3"
2와 더하기를 더하기 하면.... --> 오류 발생
```

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Splitter.kt:
```diff
@@ -0,0 +1,49 @@
+
+            var beforeNode: Node = toNode(splitList[0])
+
+            if (beforeNode is Operator) {
+                throw IllegalArgumentException("첫번째에는 연산자가 들어올 수 없습니다.")
+            }
+
+            val nodeQueue: Queue<Node> = LinkedList()
+            nodeQueue.offer(beforeNode)
+
+            for (i in CHECKING_START_INDEX until splitList.size) {
+                val node = toNode(splitList[i])
+                if (node is Operand && beforeNode !is Operator) {
+                    throw IllegalArgumentException("피연산자 이전 항목은 연산자여야만 합니다.")
+                } else if (node is Operator && beforeNode !is Operand) {
+                    throw IllegalArgumentException("연산자 이전 항목은 피연산자이어야만 합니다.")
+                } else {
+                    nodeQueue.offer(node)
+                    beforeNode = node
+                }
+            }
+
+            if (nodeQueue.last() is Operator) throw IllegalArgumentException("마지막에는 연산자가 들어올 수 없습니다.")
```
연산자의 처음, 끝, 그리고 앞뒤 관계까지 검사를 하는 시도 좋습니다. 👍🏻 

그런데 이 로직은 `Splitter`가 해야할 역할일까요?
아니면 계산기가 분할된 노드들을 계산하는 과정에서 판단할 역할일까요?

laco-dev commented on domain/src/test/java/edu/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,51 @@
+
+    @Test(expected = IllegalArgumentException::class)
+    fun 사칙연산_기호_외에_다른_문자가_들어오면_예외를_발생시킨다() {
+        calculator.evaluate("2 * 3 ) 2")
```
Junit4를 이용한다면 expectd로 예외를 예측하는 것 보다
`assertThrows(...) { ... }` 기능을 이용하는 것이 좋습니다.

사용법을 한번 찾아보고 적용 해보세요

laco-dev reviewed (changes requested):
3단계 미션 진행 하시느라 고생 많으셨습니다.

지금은 미션의 모든 기능이 하나의 커밋에 포함되어 있는데요.
구현하시는 과정에서 기능을 가장 작은 단위부터 만들어 보는 연습을 해보시면 좋습니다.
미션의 구현도 중요하지만, 구현하는 과정이 더 중요하다고 생각됩니다.
```
더하기 기능을 구현한다.
입력 값을 나누는 기능을 구현한다
```

객체지향 개념은 아직 익숙하지 않으신데도 불구하고 다른 코드 잘 참고하면서 노력하신 모습이 보기 좋았습니다.

다만 목표를 너무 깊이있게 만드려고 이것저것 개념이 끌어모아져서 코드가 상당히 복잡해 지게 되었는데, 우선은 단순하게 구현을 목표 해보시고 익숙해지신 다음 여러가지 제약사항을 확장 해 보시는것을 권장드립니다.

남겨드린 피드백은, 충분히 고민 해 보시고 도전해보세요.
추가로, 전체적으로 코드 리포매팅이 필요 해보입니다. 🙏🏻

KimKwangYoung committed:
중복된 edu경로 제거

KimKwangYoung committed:
Mock 관련 의존성 제거

KimKwangYoung committed:
Test 어노테이션 expected 옵션 제거 -> assertThrows 변경

KimKwangYoung committed:
중복된 edu경로 제거(2)

KimKwangYoung committed:
SplitterTest 추가

KimKwangYoung committed:
Splitter 리팩토링

- Calculator.evaluate()에서 유효성 검사를 진행할 수 있도록 기존 식의 유효성 검사를 하던 코드 삭제
- companion object -> object
- Queue를 쓸 이유가 없어 List로 변경

KimKwangYoung committed:
Node 리팩토링

- Node를 interface에서 sealed class로 변경
- 그에 따라 Operator도 enum class에서 data class로 변경

KimKwangYoung committed:
Calculator 리팩토링

- Splitter에서 진행하던 식의 유효성검사를 evaluate 메서드에서 진행
- throw Exception 대신 require 메서드 사용

KimKwangYoung commented:
리뷰 남겨주신거 토대로 고민 후 코드를 수정해서 다시 리뷰요청 드립니다.
피드백 주신걸 토대로 제 나름 코드를 수정하니 Splitter에서 유효성 검사를 난잡하게 하던 부분을 훨씬 줄일 수 있었던 거 같습니다.

다만 Splitter에 대한 테스트 코드를 만들어보라고 하셨는데 식이 정확하게 잘 나뉘는지 확인할 테스트 코드밖에 넣을 게 생각이 나지 않네요
테스트 코드 작성이 익숙치 않아서 이게 맞나 싶습니다😅

laco-dev commented on domain/src/test/java/edu/nextstep/camp/calculator/domain/SplitterTest.kt:
```diff
@@ -0,0 +1,14 @@
+        val expression = "5 + 3 / 4"
+        val expected = 5
+
+        assertEquals(Splitter.split(expression).size, expected)
```
> 다만 Splitter에 대한 테스트 코드를 만들어보라고 하셨는데 식이 정확하게 잘 나뉘는지 확인할 테스트 코드밖에 넣을 게 생각이 나지 않네요
테스트 코드 작성이 익숙치 않아서 이게 맞나 싶습니다😅

결과적으로 Splitter가 하는 역할이 위 테스트와 같이 여러개의 토큰으로 나누는 것이니까요.
길이를 측정하는 것도 방법이라고 봅니다.

보다 정확하게 하려면 `token[0] == "5"` 형태로 값을 비교하는 것도 있을 것 같네요

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,24 +1,28 @@
-                    operator = node
-                }
+
+    fun evaluate(inputExpression: String?):Int {
```
> 리포매팅이라고 하시면 코드에 전체적인 수정이 필요하다는 말씀이실까요?
아니면 evaluate에 대한 수정이 필요하다는 말씀이실까요?

띄어쓰기, 개행, 불필요한 import가 포함된 여부 등 모든 것을 포함합니다.
IDE의 오토 리포매팅 기능을 이용해서 코드 정리해주시면 됩니다.

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,24 +1,28 @@
+
+        val firstNode = expressionNodeList.first()
+
+        if (expressionNodeList.size % 2 != 1 || firstNode is Operator)
+            throw IllegalArgumentException("정상적인 식이 아닙니다.")
```
한 줄로 표현할 수 없는 if문의 경우 중괄호를 사용해야 합니다.
https://developer.android.com/kotlin/style-guide?hl=ko#braces
```kotlin
if (string.isEmpty())
    return  // WRONG!

if (string.isEmpty()) {
    return  // Okay
}
```

laco-dev reviewed (changes requested):
피드백 잘 반영 해주셨네요. 👍🏻 

4단계 미션에 집중하기 위해서, 코드변경이 일어날 수 있는 부분을 먼저 해결하고 넘어가면 좋을 것 같습니다.

간단한 의견 몇가지 남겨드렸으니 확인 해주세요. 🙏🏻

KimKwangYoung committed:
기존 List의 size 비교에서 node가 정상적으로 들어갔는 지 확인하는 방식으로 변경

KimKwangYoung committed:
IDE 제공 오토 리포매팅 기능을 사용하여 코드 리포매팅

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Node.kt:
```diff
@@ -0,0 +1,19 @@
+
+sealed class Node
+
+data class Operator(val symbol: String) : Node() {
+    fun calculate(left: Int, right: Int): Int {
+        return when (symbol) {
+            "+" -> left + right
+            "-" -> left - right
+            "*" -> left * right
+            "/" -> left / right
+            else -> throw IllegalArgumentException("+,-,*,/ 외에 다른 문자는 허용되지 않습니다. -> $symbol")
+        }
+    }
+}
```
이제 확인했는데, enum class에서 data class가 된 이유가 sealed clsas 때문일까요??
그렇다면, sealed interface를 사용 해 볼 수 있습니다.

사칙연산만 제공하는데, data class로 만들어질 이유가 없어 보이네요. 🙏🏻

laco-dev reviewed (approved):
피드백 반영 확인 했습니다.

아직 아래 피드백이 반영이 되지 않았는데요.
추가로 남겨드린 의견과 미반영된 내용은 다음 미션에서 같이 확인 해 주시면 됩니다. 🙏🏻 
https://github.com/next-step/android-calculator-mvc/pull/160#discussion_r922921361

3단계 미션은 종료하겠습니다. 
고생하셨습니다
